docs/ADR-8-Use-RIO-in-cardano‐cli.md

```diff
@@ -0,0 +1,132 @@
+```
+main :: IO ()
+main = toplevelExceptionHandler $ do
+  envCli <- getEnvCli
+  co <- Opt.customExecParser pref (opts envCli)
+  orDie (docToText . renderClientCommandError) $ runClientCommand co
+  
+...
+
+runClientCommand :: ClientCommand -> ExceptT ClientCommandErrors IO ()
+runClientCommand = \case
+  AnyEraCommand cmds ->
+    firstExceptT (CmdError (renderAnyEraCommand cmds)) $ runAnyEraCommand cmds
+  AddressCommand cmds ->
+    firstExceptT AddressCmdError $ runAddressCmds cmds
+  NodeCommands cmds ->
+    runNodeCmds cmds
+      & firstExceptT NodeCmdError
+  ByronCommand cmds ->
+    firstExceptT ByronClientError $ runByronClientCommand cmds
+  CompatibleCommands cmd ->
+    firstExceptT (BackwardCompatibleError (renderAnyCompatibleCommand cmd)) $
+      runAnyCompatibleCommand cmd
+...
+```
+- As a result we have a lot of errors wrapped in errors which makes the code unwieldly and difficult to compose with other code blocks. See image below of incidences of poor composability where we use `firstExceptT` (and sometimes `first`) to wrap errors in other errors.
+
+![image](https://github.com/user-attachments/assets/6a50e8f2-9140-4b7e-afa5-ff778d696742)
+ 
+- The ExceptT IO is a known anti-pattern for these reasons and others as per:
+  - https://github.com/haskell-effectful/effectful/blob/master/transformers.md
+  - https://tech.fpcomplete.com/blog/2016/11/exceptions-best-practices-haskell/
+
+
+# Proposed Solution
+
+I propose to replace  `ExceptT someError IO a` with [RIO env a](https://hackage.haskell.org/package/rio-0.1.22.0/docs/RIO.html#t:RIO).
+Below is a summary of pros and cons of `RIO` vs our current `ExceptT someError IO a` pattern.
+
+### Pros 
+- Additional [logging functionality](https://hackage.haskell.org/package/rio-0.1.22.0/docs/RIO.html#g:5)
+- Explicit environment dependencies e.g `logError :: HasLogFunc env => Text -> RIO env ()`
+- Better composability i.e no more errors that wrap errors (see above).
+
+### Cons
+- `RIO` is hardcoded to IO so we cannot add additional transformer layers e.g `RIO Env (StateT Int IO) a`
+- Implicit error flow. Errors are thrown via GHC exceptions in `IO`. See exception handling below.
+
+## Exception handling
+
+### Exception type
+```haskell
+data CustomCliException where
+  CustomCliException
+    :: (Show error, Typeable error, Error error, HasCallStack)
+    => error -> CustomCliException
+
+deriving instance Show CustomCliException
+
+instance Exception CustomCliException where
+  displayException (CustomCliException e) =
+    unlines
+      [ show (prettyError e)
+      , prettyCallStack callStack
+      ]
+```
+
+The purpose of `CustomCliException` is to represent explicitly thrown, structured errors that are meaningful to our application. 
+For other exceptions (e.g., runtime errors or unexpected failures), we rely on the `Nothing` case in our proposed top level
+handler below:
+
+### Top level handler 
+```haskell
+executeRio :: RIO () () -> IO ()
```

Comment:
I would hide the `env` parameter of the `RIO` now that it is easy, unless we are sure it is going to remain `()`, and even if we are. Maybe with a type synonym. Something like: `type CIO = RIO ()`. That way if we eventually want to add an environment we won't have to change every type signature

Reply:
Something like `type EmptyEnvironment = ()`? We will eventually use it in logging.

Reply:
I wouldn't bother with hiding it. I think we can just write function types for any environment e.g. `RIO e a`. 

I'm wondering if maybe we could put stuff like `LocalNodeConnectInfo` for online commands, or environment variables into environment. So an online command using environment variables would have type like for example:
```haskell
doSomeCliCommand 
  :: Has LocalNodeConnectInfo e
  => Has EnvCli e
  => HasCallStack
  => Foo -> Bar -> RIO e ()

class Has a e where
  obtain :: a -> e
```
But that's out of scope for this ADR.


Besides that, maybe it be useful to have a type synonym providing `HasCallStack` everywhere? e.g.

```haskell
-- CIO, like Cardano IO 
type CIO e a = HasCallStack => RIO e a
```

Reply:
> I'm wondering if maybe we could put stuff like LocalNodeConnectInfo for online commands, or environment variables into environment.

100%. That is the point of the `Has` classes :+1: .

> besides that, maybe it be useful to have a type synonym providing HasCallStack everywhere?

Good suggestion, yes we should do this.

Reply:
Ah, so we have traits. That works too